src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
 
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    stateHistory.push(currentState);
```

Comment:
It is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after the current action is processed. This ensures that the same object is not referenced multiple times in the array if no changes are made in a loop iteration. [DONT REPEAT YOURSELF]